precompiles/assets/types.go

```diff
@@ -32,27 +36,121 @@ func (p Precompile) GetDepositWithdrawParamsFromInputs(ctx sdk.Context, args []i
 	// the length of client chain address inputted by caller is 32, so we need to check the length and remove the padding according to the actual length.
 	assetAddr, ok := args[1].([]byte)
 	if !ok || assetAddr == nil {
-		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 1, "[]byte", assetAddr)
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 1, reflect.TypeOf(args[1]), args[1])
 	}
 	if len(assetAddr) != types.GeneralAssetsAddrLength {
-		return nil, xerrors.Errorf(exocmn.ErrInputClientChainAddrLength, len(assetAddr), types.GeneralClientChainAddrLength)
+		return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, len(assetAddr), types.GeneralClientChainAddrLength)
 	}
 	depositWithdrawParams.AssetsAddress = assetAddr[:clientChainAddrLength]
 
 	stakerAddr, ok := args[2].([]byte)
 	if !ok || stakerAddr == nil {
-		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, "[]byte", stakerAddr)
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, reflect.TypeOf(args[2]), args[2])
 	}
 	if len(stakerAddr) != types.GeneralClientChainAddrLength {
-		return nil, xerrors.Errorf(exocmn.ErrInputClientChainAddrLength, len(assetAddr), types.GeneralClientChainAddrLength)
+		return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, len(assetAddr), types.GeneralClientChainAddrLength)
 	}
 	depositWithdrawParams.StakerAddress = stakerAddr[:clientChainAddrLength]
 
 	opAmount, ok := args[3].(*big.Int)
 	if !ok || opAmount == nil || opAmount.Cmp(big.NewInt(0)) == 0 {
-		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, "*big.Int", opAmount)
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 3, reflect.TypeOf(args[3]), args[3])
 	}
 	depositWithdrawParams.OpAmount = sdkmath.NewIntFromBigInt(opAmount)
 
 	return depositWithdrawParams, nil
 }
+
+func (p Precompile) ClientChainInfoFromInputs(_ sdk.Context, args []interface{}) (*types.ClientChainInfo, error) {
+	if len(args) != 5 {
+		return nil, xerrors.Errorf(cmn.ErrInvalidNumberOfArgs, 5, len(args))
+	}
+	clientChain := types.ClientChainInfo{}
+	clientChainID, ok := args[0].(uint32)
+	if !ok {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 0, reflect.TypeOf(args[0]), args[0])
+	}
+	clientChain.LayerZeroChainID = uint64(clientChainID)
+
+	addressLength, ok := args[1].(uint32)
+	if !ok || addressLength == 0 {
+		return nil, fmt.Errorf(exocmn.ErrContractInputParaOrType, 1, reflect.TypeOf(args[1]), args[1])
+	}
+	if addressLength > types.GeneralAssetsAddrLength {
+		return nil, xerrors.Errorf(exocmn.ErrInvalidAddrLength, addressLength, "not greater than 32")
+	}
+	clientChain.AddressLength = addressLength
+
+	name, ok := args[2].(string)
+	if !ok || name == "" {
+		return nil, xerrors.Errorf(exocmn.ErrContractInputParaOrType, 2, reflect.TypeOf(args[2]), args[2])
+	}
+	clientChain.Name = name
+
+	metaInfo, ok := args[3].(string)
+	if !ok {
```

Comment:
Validate length of metaInfo non-zero ? We can look at setting a maximum size for the lengths later

Reply:
The initial version had a validation for this. I removed it to keep it consistent because it isn't validated in the genesis validation and this field had been set in local_node.sh. It's better to set a maximum size. I can add a maximum and the validation in both the precompile and the genesis validation.

Reply:
Additionally, generate.js might need to be changed slightly after this PR is merged.

Reply:
> It's better to set a maximum size. I can add a maximum and the validation in both the precompile and the genesis validation.

Can you create an issue for tracking this?

Reply:
I wonder if we could allow this `metaInfo` being empty? since it is extra data that is used for arbitrary data like an url or plain text. I don't check this field to be non-empty in the contract implementation.

Reply:
The `metaInfo` isn't allowed to be empty, the check was added in the last commit. It seems like the check should be implemented in the front end to avoid invalid transactions. As for the contract, would It be enough to check it in the precompile implementation?

Reply:
From design perspective, I think your question actually means: whether native module should totally trust in `ExocoreGateway` or native module should do defensive parameter validation regardless the implementation of `ExocoreGateway`? personally I prefer the native module not totally trusting the implementation of `ExocoreGateway` and check for params even if they might have been checked in `ExocoreGateway`

Reply:
Yeah, the params will be checked in the native module no matter how it's implemented in `ExocoreGateway`.